### PR TITLE
msg/async: update refcount and perf counter properly

### PR DESCRIPTION
**Pull request by majianpeng**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
retest this please

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
@tchaikov . ping.

File: src/msg/async/AsyncMessenger.cc [hunk 1]
```diff
@@ -201,6 +201,7 @@ void Processor::accept()
 	accept_error_num = 0;
 	continue;
       } else {
+	--w->references;
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.cc]: should not decrement refcount if "r == -EINTR".

File: src/msg/async/AsyncMessenger.h [hunk 2]
```diff
@@ -397,7 +397,7 @@ class AsyncMessenger : public SimplePolicyMessenger {
    */
   void unregister_conn(const AsyncConnectionRef& conn) {
     std::lock_guard l{deleted_lock};
-    if (!accepting_conns.count(conn))
+    if (!accepting_conns.count(conn) || (anon_conns.find(conn) != anon_conns.end()))
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.h]:
```suggestion
    if (!accepting_conns.count(conn) || !anon_conns.count(conn))
```

File: src/msg/async/AsyncMessenger.cc (hunk 1 above)
Comment (majianpeng) [reply src/msg/async/AsyncMessenger.cc]: r == -EINTR, w->reference still ++. why not decrement?

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
@tchaikov . which problem did cause make check always failed?

**Comment by majianpeng** [comment]
retest this please

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by majianpeng** [comment]
jenkins test make check

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
jenkins test make check

**Comment by majianpeng** [comment]
retest this please

File: src/msg/async/AsyncMessenger.h [hunk 3]
```diff
@@ -397,7 +397,7 @@ class AsyncMessenger : public SimplePolicyMessenger {
    */
   void unregister_conn(const AsyncConnectionRef& conn) {
     std::lock_guard l{deleted_lock};
-    if (!accepting_conns.count(conn))
+    if (!accepting_conns.count(conn) || (anon_conns.count(conn))
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.h]: @majianpeng  mind remove the parentheses around `anon_conns.count(conn)` ?

**Comment by tchaikov** [comment]
retest this please

File: src/msg/async/AsyncMessenger.h (hunk 3 above)
Comment (majianpeng) [reply src/msg/async/AsyncMessenger.h]: @tchaikov . why remove?

Comment (tchaikov) [reply src/msg/async/AsyncMessenger.h]: there is no need to put parentheses around this expression. why would you want to add them?

Comment (majianpeng) [reply src/msg/async/AsyncMessenger.h]: in all in one ceph cluster, fio+librbd end up, the perfcounter(active_connections) didn't decrease.

Comment (tchaikov) [reply src/msg/async/AsyncMessenger.h]:
```suggestion
    if (!accepting_conns.count(conn) || anon_conns.count(conn)
```

is what i suggest.

**Comment by majianpeng** [comment]
retest this please

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-01-07_01:48:07-rados-wip-kefu-testing-2020-01-07-0032-distro-basic-smithi/

failures fixed / tracked by

- https://github.com/ceph/ceph/pull/32517
- https://github.com/ceph/ceph/pull/32518
- https://github.com/ceph/ceph/pull/32519
- https://tracker.ceph.com/issues/43431
- https://tracker.ceph.com/issues/43492
- https://tracker.ceph.com/issues/43494

**Pull request merged**